src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using System.Linq;
+using System;
+
+namespace Bicep.Core.TypeSystem
+{
+    /// <summary>
+    /// Visitor used to collect errors caused by property assignments to run-time values when that is not allowed
+    /// </summary>
+    public sealed class DeployTimeConstantVisitor : SyntaxVisitor
+    {
+
+        private readonly SemanticModel model;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        // Since we collect errors top down, we will collect and overwrite this variable and emit it in the end
+        private SyntaxBase? errorSyntax;
+        private string? currentProperty;
+        private ObjectType? bodyObj;
+        private ObjectType? referencedBodyObj;
+
+
+        private DeployTimeConstantVisitor(SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+        }
+
+        // entry point for this visitor. We only iterate through modules and resources
+        public static void ValidateDeployTimeConstants(SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            var deploymentTimeConstantVisitor = new DeployTimeConstantVisitor(model, diagnosticWriter);
+            foreach (var declaredSymbol in model.Root.ResourceDeclarations)
+            {
+                deploymentTimeConstantVisitor.Visit(declaredSymbol.DeclaringSyntax);
+            }
+            foreach (var declaredSymbol in model.Root.ModuleDeclarations)
+            {
+                deploymentTimeConstantVisitor.Visit(declaredSymbol.DeclaringSyntax);
+            }
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            if (model.GetSymbolInfo(syntax) is ResourceSymbol resourceSymbol &&
+                resourceSymbol.Type is ResourceType resourceType &&
+                resourceType.Body is ObjectType bodyObj)
+            {
+                this.bodyObj = bodyObj;
+            }
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.bodyObj = null;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            if (model.GetSymbolInfo(syntax) is ModuleSymbol moduleSymbol &&
+                moduleSymbol.Type is ModuleType moduleType &&
+                moduleType.Body is ObjectType bodyObj)
+            {
+                this.bodyObj = bodyObj;
+            }
+            base.VisitModuleDeclarationSyntax(syntax);
+            // reset both the current object and referenced object's bodyObj
```

Comment:
Why do you need this reset for module, but not resource syntax?

Reply:
Ack! Rookie mistake.